applications/fishing-map/src/features/timebar/Timebar.tsx

```diff
@@ -39,9 +53,93 @@ const TimebarWrapper = () => {
     },
     [setBookmark]
   )
+  const mapInstance = useMapboxInstance()
+  const tilesLoading = useTilesState(mapInstance).loading
+  const currentTimeChunkId = useCurrentTimeChunkId()
+  const mapStyle = useMapStyle()
 
-  if (!start || !end) return null
+  const [stackedActivity, setStackedActivity] = useState<any>()
+  useEffect(() => {
+    if (
+      !mapInstance ||
+      !mapStyle ||
+      tilesLoading ||
+      timebarVisualisation !== TimebarVisualisations.Heatmap
+    )
+      return
+
+    const temporalgrid = mapStyle.metadata?.temporalgrid
+    if (!temporalgrid) return
+    const numSublayers = temporalgrid.numSublayers
+    const timeChunks = temporalgrid.timeChunks as TimeChunks
+    const activeTimeChunk = timeChunks?.chunks.find((c: any) => c.active) as TimeChunk
+    const chunkQuantizeOffset = activeTimeChunk.quantizeOffset
+    const numChunkFrames = activeTimeChunk.framesDelta
+
+    let n = 0
+    n = performance.now()
+
+    // Getting features within viewport - it's somehow faster to use querySource with a crude viewport filter, than using queryRendered
+    const [boundsSW, boundsNE] = mapInstance.getBounds().toArray()
+    const allFeaturesWithStyle = mapInstance
+      .querySourceFeatures(currentTimeChunkId, {
+        sourceLayer: 'temporalgrid_interactive',
+      })
+      .filter((f) => {
```

Comment:
Have you tried to use the [internal filter method](https://docs.mapbox.com/mapbox-gl-js/api/map/#map#querysourcefeatures) ? 

I guess it will have better performance as the number of features you have to manage into the component is smaller.

Reply:
as we discussed this works better, thanks for the explanation!